lib/storage/metadata/mongoclient/MongoClientInterface.js

```diff
@@ -1295,25 +1341,88 @@ class MongoClientInterface {
      */
     deleteObjectNoVer(c, bucketName, objName, params, log, cb) {
         const masterKey = formatMasterKey(objName, params.vFormat);
-        c.findOneAndDelete({
-            _id: masterKey,
-        }, {}, (err, result) => {
+        this.internalDeleteObject(c, bucketName, masterKey, {}, log, (err, res) => {
             if (err) {
+                // Should not return an error when no object is found
+                if (err.is.NoSuchKey) {
+                    return cb(null);
+                }
                 log.error(
                     'deleteObjectNoVer: error deleting object with no version',
-                    { error: err.message });
+                    { error: err.message, bucket: bucketName, key: objName });
                 return cb(errors.InternalError);
             }
-            if (result.ok !== 1) {
+            if (res.ok !== 1) {
                 log.error(
-                    'deleteObjectNoVer: failed deleting object with no version',
-                    { error: err.message });
+                    'deleteObjectNoVer: failed to delete object',
+                    { bucket: bucketName, object: objName });
                 return cb(errors.InternalError);
             }
             return cb(null);
         });
     }
 
+    /**
+     * Flags the object before deleting it, this is done
+     * to keep object metadata in the oplog, as oplog delete
+     * events don't contain any object metadata
+     * @param {Object} collection MongoDB collection
+     * @param {string} bucketName bucket name
+     * @param {string} key Key of the object to delete
+     * @param {object} filter additional query filters
+     * @param {Logger}log logger instance
+     * @param {Function} cb callback containing error
+     * and BulkWriteResult
+     * @return {undefined}
+     */
+    internalDeleteObject(collection, bucketName, key, filter, log, cb) {
+        // filter used when finding and updating object
+        const findUpdateFilter = Object.assign({
+            _id: key,
```

Comment:
should check 'value.deleted' as well, as a safety measure?

Reply:
+1